cpmpy/solvers/exact.py

```diff
@@ -0,0 +1,488 @@
+            are applied in `transform()`.
+
+            Solvers can raise 'NotImplementedError' for any constraint not supported after transformation
+        """
+        from exact import Exact as xct
+
+        # Comparisons: only numeric ones as 'only_bv_implies()' has removed the '==' reification for Boolean expressions
+        # numexpr `comp` bvar|const
+        if isinstance(cpm_expr, Comparison):
+            lhs, rhs = cpm_expr.args
+            xct_coefs, xct_vars, xct_rhs = self._make_numexpr(lhs,rhs)
+
+            # Thanks to `only_numexpr_equality()` only supported comparisons should remain
+            if cpm_expr.name == '<=':
+                return self._add_xct_constr(xct_coefs, xct_vars, False, 0, True, xct_rhs)
+            elif cpm_expr.name == '>=':
+                return self._add_xct_constr(xct_coefs, xct_vars, True, xct_rhs, False, 0)
+            elif cpm_expr.name == '==':
+                # a BoundedLinearExpression LHS, special case, like in objective
+                return self._add_xct_constr(xct_coefs, xct_vars, True, xct_rhs, True, xct_rhs)
+
+            #elif lhs.name == 'mul':
+                #assert len(lhs.args) == 2, "Exact only supports multiplication by a constant"
+                #a, b = self.solver_vars(lhs.args)
+                #assert is_num(b), "Exact only supports multiplication by a constant"
+                #return self._add_xct_constr([b], [a], True, xct_rhs, False, xct_rhs)
+
+            #elif lhs.name == 'div':
+                #assert len(lhs.args), "Exact only supports division by a constant"
+                #a, b = self.solver_vars(lhs.args)
+                #assert is_num(b), "Exact only supports division by a constant"
+                #return self._add_xct_constr([1], [a], True, xct_rhs*b, False, xct_rhs*b)
+
+            raise NotImplementedError(
+                "Constraint not supported by Exact '{}' {}".format(lhs.name, cpm_expr))
+
+        elif isinstance(cpm_expr, Operator) and cpm_expr.name == "->":
+            # Indicator constraints
+            # Take form bvar -> sum(x,y,z) >= rvar
+            cond, sub_expr = cpm_expr.args
+            assert isinstance(cond, _BoolVarImpl), f"Implication constraint {cpm_expr} must have BoolVar as lhs"
+            assert isinstance(sub_expr, Comparison), "Implication must have linear constraints on right hand side"
+
+            lhs, rhs = sub_expr.args
+            assert isinstance(lhs, _NumVarImpl) or lhs.name == "sum" or lhs.name == "wsum",f"Unknown linear expression {lhs} on right side of indicator constraint: {cpm_expr}"
+
+            xct_coefs, xct_vars, xct_rhs = self._make_numexpr(lhs,rhs)
+
+            if isinstance(cond, NegBoolView):
+                cond, bool_val = self.solver_var(cond._bv), False
+            else:
+                cond, bool_val = self.solver_var(cond), True
+
+            if sub_expr.name == "==":
```

Comment:
The design idea of our solver interfaces so far is that the transformations should transform expressions into those (almost) directly supported by the solver.

I have the feeling that here you are doing a bit too much transformation-like work. That we would better put these into a new transformation?

Reply:
Yeah, I agree. Also see https://github.com/CPMpy/cpmpy/issues/183.

Reply:
Added a comment describing the remaining transformations. We can refactor those later based on information by other solvers.